emukit/bayesian_optimization/acquisitions/local_penalization.py

```diff
@@ -0,0 +1,96 @@
+import numpy as np
+from scipy.stats import norm
+
+from emukit.core.acquisition import Acquisition
+
+
+class LocalPenalization(Acquisition):
+    """
+    Penalization function used in the local penalization method. It is the log of the following function:
+
+    .. math::
+        h(x) = \sum_i{\log\\left(\Phi\\left(\\frac{\|x - x_i\|_{2} - r_i}{s_i}\\right)\\right)}
+
+    where :math:`x_i` are the points already in the batch
+    """
+    def __init__(self, model):
+        """
+        :param model: Model
+        """
+
+        self.x_batch = None
+        self.radius = None
+        self.scale = None
+        self.model = model
+
+    @property
+    def has_gradients(self):
+        return True
+
+    def update_batches(self, x_batch, lipschitz_constant, f_min):
+        """
+        Updates the batches internally and pre-computes the parameters of the penalization function
+        """
+        self.x_batch = x_batch
+        if x_batch is not None:
+            self._compute_parameters(x_batch, lipschitz_constant, f_min)
+        else:
+            self.x_batch = None
+            self.radius = None
+            self.scale = None
+
+    def _compute_parameters(self, x_batch, lipschitz_constant, f_min):
+        """
+        Pre-computes the parameters of a penalization function
+        """
+        mean = self.model.predict(x_batch)[0]
+        variance = self.model.predict(x_batch)[1]
+
+        variance = np.maximum(1e-16, variance)
+        std_deviation = np.sqrt(variance)
+
+        # Calculate function parameters
+        radius = (mean - f_min) / lipschitz_constant
+        scale = std_deviation / lipschitz_constant
+        self.radius = radius.flatten()
+        self.scale = scale.flatten()
+        self.x_batch = x_batch
+
+    def evaluate(self, x):
+        """
+        Evaluates the penalization function value
+        """
+
+        if self.x_batch is None:
+            return np.ones((x.shape[0], 1))
+
+        distances = _squared_distance_calculation(x, self.x_batch)
+        z = (distances - self.radius) / self.scale
```

Comment:
can 'z' have a better name?

Reply:
It can indeed. `normalized_distance` I will use